keyedReducer: disable inner reducer persistence by default

jsnajdr opened the pull request:
This PR fixes an inconsistency in how reducer-wrapping functions `combineReducers` and `keyedReducer` treat the inner reducer's persistence.

`combineReducers` disables inner reducer's persistence by default:
```js
const combined = combineReducers( {
  inner: ( state = null ) => state
} );
```
disables persistence for `inner`. `serialize( combined, x )` will always return `undefined` and `deserialize( combined, x )` will always return initial state (`{ inner: null }` in our case.

One has to opt-in into persistence with `withSchemaValidation` or `withPersistence`.

`keyedReducer` behaves differently. The default is serializing the inner reducer with identity (no transform) functions:
```js
const keyed = keyedReducer( 'id', ( state = null ) => state );
```
Here `de/serialize( keyed, { a: 1 } )` will return `{ a: 1 }`.

This PR modifies `keyedReducer` to make it behave like `combineReducers`. After the patch, `serialize( keyed, { a: 1 } )` is `undefined` and `deserialize( keyed, { a: 1 } )` is `{}`.

To get the old behavior, one needs to enable persistence explicitly:
```js
const keyed = keyedReducer( 'id', withPersistence( ( state = null ) => state ) );
```

**Fixing existing usages**
Some existing `keyedReducer` usages will need an update. There are two common patterns:
```js
withSchemaValidation( schema, keyedReducer( 'id', innerReducer ) );
```
Here we want to persisting the reducer's state, and validate it against a schema that specifies the entire object, not the individual substates. To enable persistence of `innerReducer`, we need to do it explicitly:
```js
withSchemaValidation( schema, keyedReducer( 'id', withPersistence( innerReducer ) ) );
```

The second common pattern is enabling identity-based persistence with:
```js
const keyed = keyedReducer( 'id', innerReducer );
keyed.hasCustomPersistence = true;
```
The `hasCustomPersistence` flag prevents a wrapping `combineReducers` from applying `withoutPersistence` to the `keyed` reducer. This pattern gets replaced with:
```js
const keyed = keyedReducer( 'id', withPersistence( innerReducer ) );
```

The new version of `keyedReducer` always sets `hasCustomPersistence = true` on the resulting reducer and it works in all cases.

I found all reducers that need patching by grepping for files that contain both `keyedReducer` and `withSchemaValidation`, or both `keyedReducer` and `hasCustomPersistence`:
```
git grep -l keyedReducer | xargs git grep -l withSchemaValidation
git grep -l keyedReducer | xargs git grep -l hasCustomPersistence
```

jsnajdr committed:
keyedReducer: disable inner reducer persistence by default

tyxla reviewed (commented):
This is big 😅  What is the current plan for testing this thoroughly?

jsnajdr commented:
> What is the current plan for testing this thoroughly?

I'm not sure and some advice from an independent reviewer would be helpful 🙂 

In fact, this is the only part of #50222 that introduces behavior changes. Everything else involves a lot of changed code, but all of it is a straightforward refactoring that shuffles code around and doesn't change its behavior.

The main tool for testing reducer persistence are unit tests. The whole `client/state` tree is pretty well covered in that regard.

Would it help if all reducers that this PR modifies had persistence unit tests? There would be three commits:
- persistence tests for `keyedReducer` usages. They pass.
- changes to `keyedReducer` behavior. That breaks the tests.
- fixes for the `keyedReducer` usages. Mostly adding a `withPersistence` bit. That makes the tests green again.

We already have the last two commits. We don't have the first one.

The remaining question is: did we really fix all `keyedReducer` usages? I found usages that need a fix by grepping for `keyedReducer`, `withSchemaValidation` and `hasCustomPersistence`, but maybe that filter missed something.

jsnajdr committed:
Add unit tests for keyedReducer usages at risk

jsnajdr committed:
Update keyedReducer usages for new persistence rules

tyxla commented:
Cases like #50748 are what really baffles me. Did we really mean to persist or not originally? I'm concerned we might not be understanding each case's persistence end goal and we might be introducing unexpected bugs that way. Not sure how we can go through that without manually auditing every single case.

jsnajdr commented:
Rebased this on top of the already merged #50748 and added two unit tests: for `wordads/status` and `stats/counts`.

tyxla commented:
Tests seem to be failing again. Also, considering the related PRs where we fixed persistence and cleaned up, is this one ready for a review again?

jsnajdr committed:
Update chart-tabs schema to not use type shorthands

jsnajdr commented:
> Tests seem to be failing again.

That was an interesting bug in the `chart-tabs/schema.js` file. It used a shorthand for specifying property type:
```
comments: 'number'
```
vs the full
```
comments: { type: 'number' }
```
When running as unit test, the schema validation code validates the schema JSON itself against a schema for schemas 🙂 And the schema allows only the full `{ type: 'number' }` forms.

The `is-my-json-valid` library itself apparently supports the shorthands, so at runtime everything works.

> Also, considering the related PRs where we fixed persistence and cleaned up, is this one ready for a review again?

Yes it's been rebased on top of the merged AT state PR, and is now smaller again. Please have another look and let me know if there's anything suspicious or risky that deserves more testing.

tyxla commented on client/state/jetpack-connect/reducer/jetpack-auth-attempts.js:
```diff
@@ -30,9 +30,7 @@ export function authAttempts( state = undefined, { type, attemptNumber } ) {
+	keyedReducer( 'slug', withPersistence( authAttempts ) )
 );
-
-export default reducer;
```
👍

tyxla commented on client/state/posts/likes/reducer.js:
```diff
@@ -120,10 +120,7 @@ export const itemReducer = withSchemaValidation(
-postIdReducer.hasCustomPersistence = true;
-export const items = keyedReducer( 'siteId', postIdReducer );
-items.hasCustomPersistence = true;
+export const items = keyedReducer( 'siteId', keyedReducer( 'postId', itemReducer ) );
```
I'm not sure why this was like that before - was it just to disable persistence?

jsnajdr replied:
We needed to assign the intermediate reducers (like `postIdReducer`) to a variable in order to set the `hasCustomPersistence` flag on it. Now we can inline everything into one call.

A persisting reducer must have the `hasCustomPersistence` flag so that `combineReducers` doesn't ignore it. Only the `withSchemaValidation` wrapper used to add it. Other reducers needed to add it manually.

Now `keyedReducer` also adds it, correctly handling all kinds of inner reducers.

tyxla replied:
Thought so, thanks for confirming 👍

tyxla reviewed (commented):
This generally looks good. The part with post likes state was a bit odd because we were previously just specifying `hasCustomPersistence` without doing anything more.

Did we also mean to update the docs regarding `hasCustomPersistence` now that we have the persistence helper functions? It still occurs at a couple of places in inline documentation and Markdown docs.

jsnajdr commented:
> Did we also mean to update the docs regarding `hasCustomPersistence` now that we have the persistence helper functions?

I'd rather do a standalone docs update PR after we're done with the refactoring. It will be a big update 🙂

tyxla reviewed (approved):
Good to 🚀
